### PR TITLE
[EVM] implement message server that processes EVM transactions

### PR DESCRIPTION
**Pull request by codchen**
## Describe your changes and provide context
- Implement EVMMessage endpoint that instantiate an EVM instance with a `StateDBImpl` and applies it.
- Added logics for EVM transaction receipt
- Added logics to emit EVM event logs as native events

## Testing performed to validate your change
`msg_server_test`

### CODE REVIEW

File: x/evm/keeper/coinbase.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const FeeCollectorAddress = "41cc5d9842746c69d689c8379f7f5662b8701393"
+
+func (server msgServer) GetFeeCollectorAddress(ctx sdk.Context) (common.Address, error) {
```
Comment (jewei1997) [inline x/evm/keeper/coinbase.go]: hm why not call this coinbase address?

File: x/evm/keeper/coinbase.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const FeeCollectorAddress = "41cc5d9842746c69d689c8379f7f5662b8701393"
```
Comment (jewei1997) [inline x/evm/keeper/coinbase.go]: where does this come from?

File: x/evm/keeper/msg_server.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+var _ types.MsgServer = msgServer{}
+
+func (server msgServer) EVMTransaction(goCtx context.Context, msg *types.MsgEVMTransaction) (*types.MsgEVMTransactionResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	stateDB := state.NewStateDBImpl(ctx, &server)
+	tx, _ := msg.AsTransaction()
+	ctx, gp := server.getGasPool(ctx)
+	emsg, err := server.getEVMMessage(ctx, tx)
+	if err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
```
Comment (jewei1997) [inline x/evm/keeper/msg_server.go]: should we return `nil, err` here?

File: x/evm/keeper/msg_server_test.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	require.Nil(t, err)
+	bz, err = abi.Pack("set", big.NewInt(20))
+	require.Nil(t, err)
+	txData = ethtypes.LegacyTx{
+		GasPrice: big.NewInt(1),
+		Gas:      200000,
+		To:       &contractAddr,
+		Value:    big.NewInt(0),
+		Data:     bz,
+		Nonce:    1,
```
Comment (jewei1997) [inline x/evm/keeper/msg_server_test.go:84]: may be good to do a failed tx test where gas is insufficient or nonce is incorrect as well.

Comment (jewei1997) [reply x/evm/keeper/msg_server_test.go:84]: also may be good to do an EIP-1559 txn (as opposed to legacy)

File: x/evm/state/state.go [hunk 5]
```diff
@@ -31,6 +32,7 @@ func (s *StateDBImpl) getState(ctx sdk.Context, addr common.Address, hash common
 }
 
 func (s *StateDBImpl) SetState(addr common.Address, key common.Hash, val common.Hash) {
+	fmt.Println(key.Hex())
```
Comment (jewei1997) [inline x/evm/state/state.go]: nit: remove

File: x/evm/state/expected_keepers.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package state
```
Comment (jewei1997) [inline x/evm/state/expected_keepers.go:1]: curious: why is this file called expected_keepers?

File: x/evm/state/statedb.go [hunk 7]
```diff
@@ -53,5 +56,43 @@ func (s *StateDBImpl) Finalize() error {
 		s.snapshottedCtxs[i].MultiStore().(sdk.CacheMultiStore).Write()
 	}
 
+	if logs, err := s.GetLogs(); err != nil {
+		return err
+	} else {
+		for _, l := range logs {
```
Comment (jewei1997) [inline x/evm/state/statedb.go:52]: would be good to have a smart contract that emits an event to test this

**Review by jewei1997 — APPROVED** [review]
some questions and requests for tests, but overall lgtm!

File: x/evm/keeper/msg_server.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+		// infinite gas meter
+		return ctx, math.MaxUint64
+	} else {
+		return ctx, core.GasPool(ctx.BlockGasMeter().Limit() - ctx.BlockGasMeter().GasConsumedToLimit())
+	}
+}
+
+func (server msgServer) getEVMMessage(ctx sdk.Context, tx *ethtypes.Transaction) (*core.Message, error) {
+	cfg := server.GetChainConfig(ctx).EthereumConfig(server.ChainID())
+	signer := ethtypes.MakeSigner(cfg, big.NewInt(ctx.BlockHeight()), uint64(ctx.BlockTime().Unix()))
```
Comment (philipsu522) [inline x/evm/keeper/msg_server.go:86]: is this going to cause issues with clock skew? i.e. is this committed to state?

File: x/evm/keeper/msg_server.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	return core.TransactionToMessage(tx, signer, server.getBaseFee())
+}
+
+func (server msgServer) applyEVMMessage(ctx sdk.Context, msg *core.Message, stateDB vm.StateDB, gp core.GasPool) (*core.ExecutionResult, error) {
+	coinbase, err := server.GetFeeCollectorAddress(ctx)
```
Comment (philipsu522) [inline x/evm/keeper/msg_server.go:91]: why is it coinbase specific?

File: x/evm/keeper/msg_server.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+	} else if len(msg.Data) > 0 {
+		contractAddr = *msg.To
+	}
+
+	cumulativeGasUsed := res.UsedGas
+	if ctx.BlockGasMeter() != nil {
+		limit := ctx.BlockGasMeter().Limit()
+		cumulativeGasUsed += ctx.BlockGasMeter().GasConsumed()
+		if cumulativeGasUsed > limit {
+			cumulativeGasUsed = limit
```
Comment (philipsu522) [inline x/evm/keeper/msg_server.go:127]: is this expected behavior? do we want to error out or something on situations where we are past the limit?

File: x/evm/keeper/msg_server.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+		if ctx.BlockHeight() < h {
+			// future block doesn't have a hash yet
+			return common.Hash{}
+		}
+		// fetch historical hash from historical info
+		return server.getHistoricalHash(ctx, h)
+	}
+}
+
+func (server msgServer) getHistoricalHash(ctx sdk.Context, h int64) common.Hash {
```
Comment (philipsu522) [inline x/evm/keeper/msg_server.go:172]: should we return error too? what is oss evm behavior?

File: x/evm/keeper/msg_server_test.go (hunk 4 above)
Comment (philipsu522) [reply x/evm/keeper/msg_server_test.go:84]: +1, let's add unit tests for more edge cases

File: x/evm/keeper/genesis.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+)
+
+func (k Keeper) InitGenesis(ctx sdk.Context) {
```
Comment (philipsu522) [inline x/evm/keeper/genesis.go:10]: dont think it's necessary, but do we need to add migration logic?

File: x/evm/keeper/msg_server.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	res, err := server.applyEVMMessage(ctx, emsg, stateDB, gp)
+	if err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
+	}
+
+	if err := server.writeReceipt(ctx, tx, emsg, res); err != nil {
+		return &types.MsgEVMTransactionResponse{}, err
```
Comment (philipsu522) [inline x/evm/keeper/msg_server.go]: what does a writeReceipt do? Coming from very ltitle evm knowlege, some more docstrings in general (at least for each of the `new` classes, may help

File: x/evm/keeper/coinbase.go (hunk 1 above)
Comment (codchen) [reply x/evm/keeper/coinbase.go]: in Cosmos world `coinbase` is called feeCollector, but yeah it looks confusing to use them interchangeably. I'm changing every reference under `evm` to `coinbase` instead

File: x/evm/keeper/coinbase.go (hunk 2 above)
Comment (codchen) [reply x/evm/keeper/coinbase.go]: just a deterministic non-EOA address. I've changed it to be more explicit in its generation

File: x/evm/state/expected_keepers.go (hunk 6 above)
Comment (codchen) [reply x/evm/state/expected_keepers.go:1]: it's a conventional name in Cosmos codebase for keeper interfaces. Using an interface here to avoid circular dependency between `state` module and `keeper` module

File: x/evm/keeper/msg_server.go (hunk 8 above)
Comment (codchen) [reply x/evm/keeper/msg_server.go:86]: ctx.BlockTime here is the consensus timestamp included in the block proposal and is guaranteed to be always increasing

File: x/evm/keeper/msg_server.go (hunk 9 above)
Comment (codchen) [reply x/evm/keeper/msg_server.go:91]: `coinbase` here is an Ethereum term that's basically equivalent to `feeCollector` in Sei, not the exchange.

File: x/evm/keeper/msg_server.go (hunk 10 above)
Comment (codchen) [reply x/evm/keeper/msg_server.go:127]: it will eventually cause a panic here https://github.com/sei-protocol/sei-cosmos/blob/main/baseapp/baseapp.go#L1006C12-L1006C12. Here we want to still write cumulative gas to receipt (even if it will be reverted when the panic happens) because there might be other messages in the same tx (e.g. if the evm message is sent by a CW contract) that depend on this receipt

File: x/evm/keeper/msg_server.go (hunk 11 above)
Comment (codchen) [reply x/evm/keeper/msg_server.go:172]: in ethereum it returns empty if block is not found https://github.com/ethereum/go-ethereum/blob/master/core/evm.go#L88

File: x/evm/keeper/genesis.go (hunk 12 above)
Comment (codchen) [reply x/evm/keeper/genesis.go:10]: we will need migration logic to add this new module, but no custom state mutation logics are needed

File: x/evm/keeper/msg_server_test.go (hunk 4 above)
Comment (codchen) [reply x/evm/keeper/msg_server_test.go:84]: added both

File: x/evm/state/statedb.go (hunk 7 above)
Comment (codchen) [reply x/evm/state/statedb.go:52]: modified SimpleStorage to also emit an event and checked it in the first test in msg_server_test

File: x/evm/keeper/msg_server.go (hunk 13 above)
Comment (codchen) [reply x/evm/keeper/msg_server.go]: yeah, added in the keeper files
